Take2 on config unification

costinm opened the pull request:
Changes: 
- move values.yaml to the normal place ( so updateValues is not needed for normal operation, --set in helm
is more natural )
- added few targets to test generation of istio.yaml from helm
- use the same names as in the (release) istio.yaml - we want to be able to gracefully upgrade
- removed some of the configuration that is not supported ( like arbitrary names )
- use the single repo/tag - we don't really test or support getting each piece at a different version,
and the UX is pretty bad (user having to type the tag 7 times)

I did more manual diffs - they're very time consuming, fixed few things missing (ingress certs), but
we'll need more people and eyes to get them in sync and make sure nothing gets lost.

istio-testing commented:
@costinm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 2a73be5bce59da879911d93e104accabd4efec5c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-pilot-e2e/3186/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

sdake commented on install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml:
```diff
@@ -38,7 +38,7 @@ spec:
             mountPath: /etc/statsd
         - name: {{ template "mixer.name" . }}
-          image: "{{ .Values.global.mixer.repository }}:{{ .Values.global.mixer.tag }}"
+          image: "{{ .Values.global.hub }}/mixer:{{ .Values.global.tag }}"
```
this is a good improvement.

sdake commented on install/kubernetes/helm/istio/charts/ingress/templates/deployment.yaml:
```diff
@@ -33,7 +31,7 @@ spec:
           - "2"
           - --discoveryAddress
-          - {{ .Release.Name }}-pilot:15003
+          - istio-pilot:15003
```
yup this is much better.

sdake commented on install/kubernetes/helm/istio/charts/ingress/templates/service.yaml:
```diff
@@ -1,28 +1,43 @@
     heritage: {{ .Release.Service }}
     istio: ingress
 spec:
-{{- if .Values.service.nodePort.enabled }}
```
the nodePort.enabled feature is to permit people to eval if they lack a load balancer.  I think the current istio generated manifest has this nodeport feature commented out.  I think the conditional is warranted.

costinm replied:
Super useful for minikube - but instead of conditional I added 2 services, one with node port 
and one load balancer. This way we can have both.

sdake replied:
wfm.

sdake commented on install/kubernetes/helm/istio/charts/ingress/templates/service.yaml:
```diff
@@ -1,28 +1,43 @@
+    istio: ingress
+
+---
+{{- if .Values.service.nodePort.enabled }}
```
cool if this works, wfm.

sdake commented on install/kubernetes/helm/istio/values.yaml:
```diff
@@ -1,38 +1,38 @@
+# Common settings.
 global:
+  # Default tag for Istio images.
+  hub: istio
```
is this hub correct?  Wouldn't it be docker.io/istio or gcr.io/istio?

costinm replied:
docker.io is added automatically, and it seems to work.

sdake replied:
wfm.

linsun replied:
would be good to specify it out as I'm not sure it'd work when custom registry is used in various cloud provider.  also for clarity :)

costinm replied:
Will do it in a separate PR, more changes will likely be needed before we change the tests.

costinm commented on install/kubernetes/helm/istio/charts/servicegraph/templates/deployment.yaml:
```diff
@@ -12,27 +12,26 @@ spec:
-            httpGet:
-              path: /
-              port: {{ .Values.service.internalPort }}
+#          livenessProbe:
```
commented out because it fails - 404. We need to add it back, but not sure what URL.

sdake commented:
When I tested, I accidentally deployed to the istio namespace rather than the istio-system namespace which results in:
```sdake@falkor-07:~/go/src/istio.io/istio/samples/bookinfo/kube$ istioctl kube-inject -i istio -f bookinfo.yaml -o bookfino-injected.yaml
Error: failed to convert to proto. unknown field "mixerCheckServer" in v1alpha1.MeshConfig
sdake@falkor-07:~/go/src/istio.io/istio/samples/bookinfo/kube$ istioctl version
Version: 0.5.1
GitRevision: 30acfe6528107ea333543309095659b93364b30d
User: root@2e4a18076b04
Hub: docker.io/istio
GolangVersion: go1.9
BuildStatus: Clean```

I'll try istio-system.  In the past, the helm charts have always deployed by default to the default namespace.  IOW there may be a latent issue with the helm chart related to custom namespaces.  Otherwise, things look good!

costinm commented:
Interesting point - we typically recommend using istio-system, and any
upgrade from the existing
template to the new ones would go there. The existing template has explicit
'namespace' statements,
I should probably add them as a conditional.

At the same time - there is no reason the template shouldn't work in the
default or other namespaces
(except it's not a recommended config). I suspect it's a 0.5.1 bug - I
tested with a '--set' option pointed
to master.

On Fri, Feb 23, 2018 at 7:01 AM, Steven Dake <notifications@github.com>
wrote:

> When I tested, I accidentally deployed to the istio namespace rather than
> the istio-system namespace which results in:
>
> Error: failed to convert to proto. unknown field "mixerCheckServer" in v1alpha1.MeshConfig
> sdake@falkor-07:~/go/src/istio.io/istio/samples/bookinfo/kube$ istioctl version
> Version: 0.5.1
> GitRevision: 30acfe6528107ea333543309095659b93364b30d
> User: root@2e4a18076b04
> Hub: docker.io/istio
> GolangVersion: go1.9
> BuildStatus: Clean```
>
> I'll try istio-system.  In the past, the helm charts have always deployed by default to the default namespace.  IOW there may be a latent issue with the helm chart related to custom namespaces.  Otherwise, things look good!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/3594#issuecomment-368033228>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFI6kUmEpthUWt5UbdlGSOKUCg8vMMGks5tXtLVgaJpZM4SJlMM>
> .
>

sdake commented:
Agreed it is likely a problem with the 0.51 images being out of sync with the manifests.  I have just rebuilt latest master and will report back whether that works for me or not with `helm install istio --name istio --set hub=docker.io/sdake --set tag=sdake`. 

Cheers
-steve

linsun commented on install/kubernetes/helm/istio/charts/grafana/templates/deployment.yaml:
```diff
@@ -14,14 +14,13 @@ spec:
     metadata:
       labels:
-        app: {{ template "grafana.name" . }}
-        release: {{ .Release.Name }}
```
I think there is value to keep release label so users can easily query what makes up a helm release.

costinm replied:
This breaks upgrading from istio.yaml ( release / upgradeVersion ) to istio.yaml (helm ), which is 
pretty important for now.

We can add annotations or other markers. I'm also not sure if the cost is worth it (k8s keeps indexes by label)

linsun commented on install/kubernetes/helm/istio/charts/grafana/templates/service.yaml:
```diff
@@ -15,5 +15,4 @@ spec:
       name: {{ .Values.service.name }}
   selector:
-    app: {{ template "grafana.name" . }}
-    release: {{ .Release.Name }}
```
same comment here...

costinm replied:
Confirmed that the release/etc show up in the 'deployment' (as labels). 
The change only removes them from the replica set and pods.

linsun commented on install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml:
```diff
@@ -15,13 +15,11 @@ spec:
       labels:
-        app: {{ template "mixer.name" . }}
-        release: {{ .Release.Name }}
         istio: mixer
```
I think these labels are helm convention you may want to keep.

costinm replied:
I tried. We can add them later ( but will break upgrade ). 

k8s treats the label set as immutable for some strange reasons.

linsun commented on install/kubernetes/helm/istio/charts/security/templates/service.yaml:
```diff
@@ -0,0 +1,17 @@
+  labels:
+    app: istio-ca
+spec:
+  type: LoadBalancer
```
do we always want to do loadbalancer type for this?  I asked as istio ingress has a node port option.

costinm replied:
That's what we have in istio.yaml - we can add a node port as well, but separate PR :-)

linsun commented on install/kubernetes/helm/istio/charts/sidecar-injector/templates/service.yaml:
```diff
@@ -1,17 +1,11 @@
-    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
     istio: sidecar-injector
```
same comment here on labels.

linsun commented on install/kubernetes/helm/istio/templates/configmap.yaml:
```diff
@@ -9,7 +9,7 @@ metadata:
 data:
   mesh: |-
-  {{- if .Values.global.securityEnabled }}
+  {{- if .Values.global.mtlsDefault }}
```
could we call it mtlsEnabled?   default is confusing as to on or off...

costinm replied:
Do you mind doing this in separate PR ? Don't want to go trough another wait :-)

costinm replied:
I had to rebase, so updated it.

linsun commented on install/kubernetes/helm/istio/values.yaml:
```diff
@@ -1,38 +1,38 @@
-  # create RBAC resources
+  # Default mtls policy. If true, mtls between services will be enabled by default.
+  mtlsDefault: false
+
```
I'm confused that we want to expose both securityEnabled and mtlsDefault or mtlsEnabled to our user... because you can't have mtlsEnabled = true while securityEnabled = false.  Maybe having mtlEnabled as a child property under securityEnabled?

costinm replied:
Well, technically you can... SecurityEnabled means istioCA is installed. In theory you can have
another component generating certificates ( for example using vault or a native CA). Or use 
some other spiffee thing.

linsun commented on install/kubernetes/helm/istio/values.yaml:
```diff
@@ -1,38 +1,38 @@
+
+  # create RBAC resources. Must be set for any cluster configured with rbac.
   rbacEnabled: true
-  # mutual TLS authentication
```
for rbacEnabled... I'd assume it is not valid when security is enabled?   and mixer is enabled today...

costinm replied:
AFAIK RBAC is required if the cluster has rbac ( and many have it on by default ).
RBAC only registers the cluster permissions so pilot can read stuff.

istio-testing commented:
@costinm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5c837b03d6a19f80123c4d73bbe29b88f6892e57 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5677/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5c837b03d6a19f80123c4d73bbe29b88f6892e57 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5677/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | 5c837b03d6a19f80123c4d73bbe29b88f6892e57 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1086/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5c837b03d6a19f80123c4d73bbe29b88f6892e57 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5677/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | e241021e14b4616d160f96d59228a5891c0aee68 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1088/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | e241021e14b4616d160f96d59228a5891c0aee68 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1088/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | e241021e14b4616d160f96d59228a5891c0aee68 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5686/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | e241021e14b4616d160f96d59228a5891c0aee68 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5686/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | 7c846955385360e38d6551509625a8b0be8c4669 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1099/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 7c846955385360e38d6551509625a8b0be8c4669 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1099/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 7c846955385360e38d6551509625a8b0be8c4669 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5707/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

costinm commented:
PTAL - I addressed most of the comments, but also added prometheus change (to sync with the templates/addons config), and added the refresh setting for larger clusters.

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 7c846955385360e38d6551509625a8b0be8c4669 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5707/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | 0c10bd35c78e898f6a8234e371665c802d3eca08 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1100/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 0c10bd35c78e898f6a8234e371665c802d3eca08 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1100/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 0c10bd35c78e898f6a8234e371665c802d3eca08 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5708/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 0c10bd35c78e898f6a8234e371665c802d3eca08 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5708/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | de606e42517a5ed40c605456fe7efcbe176dd13d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1115/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | de606e42517a5ed40c605456fe7efcbe176dd13d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-unit-tests/1115/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | de606e42517a5ed40c605456fe7efcbe176dd13d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-presubmit/5731/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@costinm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | dcc0bc79a45913854f17cf9efbde6b72c65b80ee | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3594/istio-pilot-e2e/3548/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ldemailly commented on bin/init.sh:
```diff
@@ -89,6 +91,15 @@ if [ ! -f ${ISTIO_OUT}/envoy ] ; then
+if [ ! -f /usr/local/bin/helm -a ! -f ${ISTIO_OUT}/helm ] ; then
+    # Install helm. Please keep it in sync with .circleci
+    cd /tmp && \
+        curl -Lo /tmp/helm.tgz https://storage.googleapis.com/kubernetes-helm/helm-${HELM_VER}-linux-amd64.tar.gz && \
```
can we make this optional ? not everyone should have to install helm ? (only if you touch the templates,...)

not every build (lint, unit test, etc...) need helm either

costinm replied:
We plan to use helm to generate istio.yaml and test configs - so it won't be optional. 
Also helm ( full helm, with tiller ) will need to be part of one of the integration tests.

ldemailly replied:
this is a bit risky though because helm can easily become required accidentally that way
it's also not hermetic, we should avoid downloading stuff
maybe the helm client can be vendored ? it should already have most of it as we use the libs in gen-deploy ?

in short I'm concerned it's really easy after that to break the "istio-auth.yaml" is all you need to install a basic istio (demo) setup - and pass all the tests

ldemailly replied:
it still shouldn't be required

costinm commented:
/test istio-pilot-e2e
